app/models/item.rb

```diff
@@ -77,8 +77,22 @@ def deleted?
     deleted_at != nil
   end
 
-  def total_value
-    current_quantity * value
+  def version_at(at)
+    return self if at >= updated_at
+    versions.sort_by(&:created_at).reverse
+            .find { |v| v.created_at <= at }.try(:reify)
+  end
+
+  def total_value(at: nil)
+    if at.blank?
+      current_quantity * value
+    else
+      prev_version = version_at(at)
```

Comment:
This is where I first noticed the smell of unclear naming. I think rather than putting the logic in total value we should make a new method `def past_value(time)`

I could see logic for the argument in the reports or if you'd rather here how about some short circut logic
```
def total_value(time: nil)
  return current_quantity * value if at.blank? || time >= updated_at
  past_version_value(time)
end

def past_version_value(time)
  past_version = versions.sort_by(&:created_at).reverse
                         .find { |v| v.created_at <= at }.try(:reify)
  prev_version.current_quantity * prev_version.value
end
```

Reply:
Good point and suggestion! I'll work towards putting this in.